File: Sources/InstantTranslator/ViewModels/FeedbackViewModel.swift
import ControlVCore
import Foundation

@MainActor
@Observable
final class FeedbackViewModel {
    var rating: Int?
    var category: FeedbackCategory = .idea
    var message: String = ""
    var contactEmail: String

    private(set) var isSending = false
    private(set) var didSend = false
    private(set) var lastError: String?

    private let client: FeedbackClientProtocol
    private let tracker: FeedbackPromptTracker
    private let installID: String
    private let sessionToken: String?
    private let appVersion: String

    init(
        client: FeedbackClientProtocol,
        tracker: FeedbackPromptTracker,
        installID: String,
        sessionToken: String?,
        contactEmail: String?,
        appVersion: String,
        initialRating: Int? = nil
    ) {
        self.client = client
        self.tracker = tracker
        self.installID = installID
        self.sessionToken = sessionToken
        self.contactEmail = contactEmail ?? ""
        self.appVersion = appVersion
        self.rating = initialRating
    }

    /// A rating alone is enough (quick "★★★★★"); otherwise a message is required.
    var canSend: Bool {
        !isSending && (rating != nil || !trimmedMessage.isEmpty)
    }

    private var trimmedMessage: String {
        message.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func send() async {
        guard canSend else { return }
        isSending = true
        lastError = nil
        defer { isSending = false }

        let email = contactEmail.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let submission = FeedbackSubmission(
            rating: rating,
            category: category,
            message: trimmedMessage,
            contactEmail: email.contains("@") ? email : nil,
            installID: installID,
            sessionToken: sessionToken,
            appVersion: appVersion
        )

        do {
            try await client.submit(submission)
            tracker.markSubmitted()
            didSend = true
        } catch {
            lastError = error.localizedDescription
        }
    }
}

File: windows/ControlV.App/ViewModels/FeedbackViewModel.cs
using System.Net.Http;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using ControlV.Core;
using ControlV.Core.Models;

namespace ControlV.App.ViewModels;

internal sealed partial class FeedbackViewModel : ObservableObject
{
    private readonly IFeedbackClient _client;
    private readonly FeedbackPromptTracker _tracker;
    private readonly Func<string> _installId;
    private readonly Func<string?> _sessionToken;
    private readonly Func<string?> _contactEmail;

    public FeedbackViewModel(IFeedbackClient client, FeedbackPromptTracker tracker, Func<string> installId, Func<string?> sessionToken, Func<string?> contactEmail)
    { _client = client; _tracker = tracker; _installId = installId; _sessionToken = sessionToken; _contactEmail = contactEmail; }

    public IReadOnlyList<FeedbackCategory> Categories => Enum.GetValues<FeedbackCategory>();
    [ObservableProperty] private int? _rating;
    [ObservableProperty] private FeedbackCategory _category = FeedbackCategory.Idea;
    [ObservableProperty] private string _message = "";
    [ObservableProperty] private string _contactEmailInput = "";
    [ObservableProperty] private bool _isSending;
    [ObservableProperty] private bool _didSend;
    [ObservableProperty] private string? _lastError;
    public bool ShouldInvite => _tracker.ShouldInvite;
    public bool CanSend => !IsSending && (Rating is not null || !string.IsNullOrWhiteSpace(Message));

    partial void OnRatingChanged(int? value) => OnPropertyChanged(nameof(CanSend));
    partial void OnMessageChanged(string value) => OnPropertyChanged(nameof(CanSend));

    public void Reset(int? rating)
    {
        Rating = rating; Category = FeedbackCategory.Idea; Message = ""; DidSend = false; LastError = null;
        ContactEmailInput = _contactEmail() ?? "";
    }

    [RelayCommand] private void SetRating(int value) => Rating = value;
    [RelayCommand] private void Dismiss() { _tracker.MarkDismissed(); OnPropertyChanged(nameof(ShouldInvite)); }

    [RelayCommand] private async Task SendAsync()
    {
        if (!CanSend) return;
        IsSending = true; LastError = null; OnPropertyChanged(nameof(CanSend));
        try
        {
            var email = ContactEmailInput.Trim().ToLowerInvariant();
            await _client.SubmitAsync(new FeedbackSubmission
            {
                Rating = Rating, Category = Category.RawValue(), Message = Message.Trim(),
                ContactEmail = email.Contains('@') ? email : null, InstallId = _installId(), SessionToken = _sessionToken(),
                AppVersion = Services.AppPaths.AppVersion,
            });
            _tracker.MarkSubmitted(); DidSend = true; OnPropertyChanged(nameof(ShouldInvite));
        }
        catch (Exception e) when (e is AuthException or HttpRequestException) { LastError = e.Message; }
        finally { IsSending = false; OnPropertyChanged(nameof(CanSend)); }
    }
}
